File: conformance-tests/complex/query-data-ref_spec_m4.js
function getCurrent(queryDataRef) {
  const context = {};
  const queryData = queryDataRef.current || (queryDataRef.current = {
    context,
    onNewData() {}
  });

  return queryData;
}

describe('Query Data Ref', () => {
  it('should use existing current', () => {
    const givenCurrent = {
      context: 'context', onNewData() {
      }
    }
    const given = {
      current: givenCurrent
    };
    const result = getCurrent(given);

    expect(result).toBe(givenCurrent)
  });

  it('should initialize current', () => {
    const given = {};
    const result = getCurrent(given);

    expect(result).toBeDefined()
    expect(given.current).toBeDefined()
    expect(result).toBe(given.current)
  });
});

File: conformance-tests/complex/query-data-ref_spec_m4.lua
-- ROBLOX upstream: https://github.com/Roblox/js-to-lua/blob/sha/conformance-tests/complex/query-data-ref_spec_m4.js
local Packages --[[ ROBLOX comment: must define Packages module ]]
local LuauPolyfill = require(Packages.LuauPolyfill)
local Boolean = LuauPolyfill.Boolean
local function getCurrent(queryDataRef)
	local context = {}
	local queryData = Boolean.toJSBoolean(queryDataRef.current) and queryDataRef.current
		or (function()
			queryDataRef.current = { context = context, onNewData = function(self) end }
			return queryDataRef.current
		end)()
	return queryData
end
describe("Query Data Ref", function()
	it("should use existing current", function()
		local givenCurrent = { context = "context", onNewData = function(self) end }
		local given = { current = givenCurrent }
		local result = getCurrent(given)
		expect(result).toBe(givenCurrent)
	end)
	it("should initialize current", function()
		local given = {}
		local result = getCurrent(given)
		expect(result).toBeDefined()
		expect(given.current).toBeDefined()
		expect(result).toBe(given.current)
	end)
end)
